crates/containerd-shim-wasmedge/src/instance.rs

```diff
@@ -103,171 +134,158 @@ pub fn reset_stdio() {
     }
 }
 
-pub fn prepare_module(
-    mut vm: Vm,
-    spec: &oci::Spec,
-    stdin_path: String,
-    stdout_path: String,
-    stderr_path: String,
-) -> Result<Vm, WasmRuntimeError> {
-    debug!("opening rootfs");
-    let rootfs_path = oci::get_root(spec).to_str().unwrap();
-    let root = format!("/:{}", rootfs_path);
-    let mut preopens = vec![root.as_str()];
-
-    debug!("opening mounts");
-    let mut mounts = oci_utils::get_wasm_mounts(spec);
-    preopens.append(&mut mounts);
-
-    let args = oci::get_args(spec);
-    let envs = oci_utils::env_to_wasi(spec);
-
-    debug!("setting up wasi");
-    let mut wasi_instance = vm.wasi_module()?;
-    wasi_instance.initialize(
-        Some(args.iter().map(|s| s as &str).collect()),
-        Some(envs.iter().map(|s| s as &str).collect()),
-        Some(preopens),
-    );
-
-    debug!("opening stdin");
-    let stdin = maybe_open_stdio(&stdin_path).context("could not open stdin")?;
-    if stdin.is_some() {
-        unsafe {
-            STDIN_FD = Some(dup(STDIN_FILENO));
-            dup2(stdin.unwrap(), STDIN_FILENO);
-        }
-    }
+#[derive(Serialize, Deserialize)]
+struct Options {
+    root: Option<PathBuf>,
+}
 
-    debug!("opening stdout");
-    let stdout = maybe_open_stdio(&stdout_path).context("could not open stdout")?;
-    if stdout.is_some() {
-        unsafe {
-            STDOUT_FD = Some(dup(STDOUT_FILENO));
-            dup2(stdout.unwrap(), STDOUT_FILENO);
-        }
-    }
+fn determine_rootdir<P: AsRef<Path>>(bundle: P) -> Result<PathBuf, Error> {
```

Comment:
We also need to account for the containerd namespace otherwise we could end up with 2 containers with the same ID but different namespaces writing to the same state dir.

Reply:
I made it consider the namespace.
https://github.com/containerd/runwasi/pull/78/commits/3fc44d2df7a0e3d52a174d92e6316909b87c7c62